controller/src/main/java/io/pravega/controller/server/bucket/BucketService.java

```diff
@@ -0,0 +1,275 @@
+                notificationLoop.set(Futures.loop(this::isRunning, this::processNotification, executor));
+                log.info("{}: Notification loop started for bucket {}", serviceType, bucketId);
+
+                workerLoop.set(Futures.loop(this::isRunning, this::work, executor));
+                log.info("{}: Notification loop started for bucket {}", serviceType, bucketId);
+
+            } finally {
+                log.info("{}: bucket {} service start completed", getServiceType(), getBucketId());
+                serviceStartFuture.complete(null);
+            }
+        });
+    }
+
+    abstract void startBucketChangeListener();
+
+    abstract void stopBucketChangeListener();
+
+    private CompletableFuture<Void> processNotification() {
+        return notifications.take(MAX_NOTIFICATIONS_TO_TAKE).thenAccept(queue -> {
+            queue.forEach(notification -> {
+                final StreamImpl stream;
+                switch (notification.getType()) {
+                    case StreamAdded:
+                        log.info("{}: New stream {}/{} added to bucket {} ", serviceType, notification.getScope(),
+                                notification.getStream(), bucketId);
+                        stream = new StreamImpl(notification.getScope(), notification.getStream());
+                        long nextRun = System.currentTimeMillis() + executionPeriod.toMillis();
+
+                        synchronized (lock) {
+                            if (!knownStreams.contains(stream)) {
+                                knownStreams.add(stream);
+                                workQueue.add(new QueueElement(stream, nextRun));
+                            }
+                        }
+                        break;
+                    case StreamRemoved:
+                        log.info("{}: Stream {}/{} removed from bucket {}", serviceType, notification.getScope(),
+                                notification.getStream(), bucketId);
+                        stream = new StreamImpl(notification.getScope(), notification.getStream());
+                        synchronized (lock) {
+                            knownStreams.remove(stream);
+                        }
+                        break;
+                    case ConnectivityError:
+                        log.warn("{}: StreamNotification for connectivity error", serviceType);
+                        break;
+                }
+            });
+        });
+    }
+
+    private CompletableFuture<Void> work() {
+        long time = System.currentTimeMillis();
+        QueueElement element;
+        long delayInMillis = 0L;
+        synchronized (lock) {
+            if (avaiableSlots > 0) {
+                element = workQueue.peek();
+                if (element != null && element.nextExecutionTimeInMillis <= time) {
+                    // Note: we can poll on queue while holding the lock because we know the element exists.
+                    element = workQueue.poll();
```

Comment:
I see this type doesn't actually implement the queue interface. In that case, it may be worth adding a `remove` method to it with the same semantics as the one on the standard queue interface.

Reply:
```Both poll() and remove() method is used to remove head object of the Queue. The main difference lies when the Queue is empty(). If Queue is empty then poll() method will return null . ... If queue is empty then peek() method also returns null.```
So since we already peek, its the same thing for us. in fact we dont need to peek unless its a blocking queue. And `PriorityQueue` is non blocking so poll will return null immediately and we can clean this piece of code further. 
but i dont see any value in using `remove` as opposed to `poll`